### PR TITLE
[Data Explorer] Allow render from View directly, not from Data Explorer

### PR DESCRIPTION
**Pull request by ananzh**
### Description
This PR avoids re-rendering the entire AppContainer when data services update. The re-render is caused by history object which is in AppMountParameters. This history object is part of the application's routing context and can change frequently as the url is updated by query, filter or time range. Each change in the history object could potentially trigger a re-render of the AppContainer and its child components.
    
With this PR, the AppContainer will not be re-loaded. Instead, each component, like Canvas and Panel, will be re-rendered based on the updates from data.

### Issues Resolved
NA

## Screenshot
* new workflow

https://github.com/opensearch-project/OpenSearch-Dashboards/assets/79961084/ab505ca5-4045-4813-8f88-16b8df2733e1

## Changelog
- feat: [Data Explorer] Allow render from View directly, not from Data Explorer

### Check List

- [x] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by ashwin-pc** [comment]
the issue with this approach is that when the user changes the query, the current context is lost. In 2.9 the lable once rendered does not change until eother the new result comes in or there are no results

**Comment by ashwin-pc** [comment]
if history is the reaso for the rerender, cant we just move it out of the props?

**Comment by ananzh** [comment]
> the issue with this approach is that when the user changes the query, the current context is lost. In 2.9 the lable once rendered does not change until eother the new result comes in or there are no results

Let me evaluate whether Discover needs history passed from Data Explorer.

File: CHANGELOG.md [hunk 1]
```diff
@@ -54,6 +54,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - [Chrome] Introduce registerCollapsibleNavHeader to allow plugins to customize the rendering of nav menu header ([#5244](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5244))
 - [Dynamic Configurations] Pass request headers when making application config calls ([#6164](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6164))
 - [Discover] Options button to configure legacy mode and remove the top navigation option ([#6170](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6170))
+- [Performance][Discover] Remove double render ([#6167](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6167))
```
Comment (AMoo-Miki) [inline CHANGELOG.md:57]:
```suggestion
- Prevent unnecessary rendering of the Data Explorer ([#6167](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6167))
```

**Comment by kavilla** [comment]
@ananzh, double check if this still relevant and next steps.

**Comment by ananzh** [comment]
> if history is the reason for the rerender, cant we just move it out of the props?

We can't. History is used in Data explorer and it is not easy to remove.

**Review by AMoo-Miki — APPROVED** [review]

**Comment by ashwin-pc** [comment]
> the issue with this approach is that when the user changes the query, the current context is lost. In 2.9 the lable once rendered does not change until eother the new result comes in or there are no results

@ananzh did we resolve this?

**Comment by ananzh** [comment]
> > the issue with this approach is that when the user changes the query, the current context is lost. In 2.9 the lable once rendered does not change until eother the new result comes in or there are no results
> 
> @ananzh did we resolve this?

I have used a diff approach and don't have that issue. Discover can still use its own history.

**Comment by kavilla** [comment]
@ananzh I think one of the tests are failing for refresh interval after re-running still. Mind checking it?

**Review by ashwin-pc — DISMISSED** [review]

**Comment by ananzh** [comment]
> @ananzh I think one of the tests are failing for refresh interval after re-running still. Mind checking it?

@kavilla I have fixed the test in https://github.com/opensearch-project/opensearch-dashboards-functional-test/pull/1358

Since we don't have double render, the tmp solution 
```
Cypress.Commands.add('makeDatePickerMenuOpen', () => {
  cy.get(
    '[class="euiFormControlLayout euiFormControlLayout--group euiSuperDatePicker"]'
  ).then(($popover) => {
    // Check if the popover does not have the 'euiPopover-isOpen' class
    if (!$popover.hasClass('euiPopover-isOpen')) {
      // If not open, click the button to open the quick menu
      cy.getElementByTestId('superDatePickerToggleQuickMenuButton').click();
    }
  });
})
```
, which is just to re-click the DatePickerToggle to make the menu re-open, is not required. We could just simply remove this tmp fix. 


https://github.com/opensearch-project/OpenSearch-Dashboards/assets/79961084/b5551487-21b5-44e7-a4d4-d959fc372b24

File: src/plugins/data_explorer/public/utils/use/shallow_equal.ts
Comment (AMoo-Miki) [inline src/plugins/data_explorer/public/utils/use/shallow_equal.ts:1]: nit: A better home for this is @osd/std.

**Review by AMoo-Miki — APPROVED** [review]

**Review by kavilla — APPROVED** [review]

**Pull request merged**